sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -104,16 +104,48 @@ abstract class Expression extends TreeNode[Expression] {
     }.getOrElse {
       val isNull = ctx.freshName("isNull")
       val value = ctx.freshName("value")
-      val ve = doGenCode(ctx, ExprCode("", isNull, value))
-      if (ve.code.nonEmpty) {
+      val eval = doGenCode(ctx, ExprCode("", isNull, value))
+      reduceCodeSize(ctx, eval)
+      if (eval.code.nonEmpty) {
         // Add `this` in the comment.
-        ve.copy(code = s"${ctx.registerComment(this.toString)}\n" + ve.code.trim)
+        eval.copy(code = s"${ctx.registerComment(this.toString)}\n" + eval.code.trim)
       } else {
-        ve
+        eval
       }
     }
   }
 
+  private def reduceCodeSize(ctx: CodegenContext, eval: ExprCode): Unit = {
+    // TODO: support whole stage codegen too
+    if (eval.code.trim.length > 500 && ctx.INPUT_ROW != null && ctx.currentVars == null) {
+      val setIsNull = if (eval.isNull != "false" && eval.isNull != "true") {
+        val globalIsNull = ctx.freshName("globalIsNull")
+        ctx.addMutableState(ctx.JAVA_BOOLEAN, globalIsNull)
+        val localIsNull = eval.isNull
+        eval.isNull = globalIsNull
+        s"$globalIsNull = $localIsNull;"
+      } else {
+        ""
+      }
+
+      val javaType = ctx.javaType(dataType)
+      val newValue = ctx.freshName("value")
```

Comment:
why is this needed? I think we can use `eval.value` instead of it

Reply:
`ev.value` may be a global variable and here we need a local variable.

Reply:
why do we strictly need a local variable here? Can't we simply assign `ev.value` to the generated function return value?

Reply:
then how are we going to change this?
`eval.code = s"$javaType $newValue = $funcFullName(${ctx.INPUT_ROW});"`

Saving a local variable is nothing and I think we shouldn't complicate the code(check if a variable is global) because of this.

Reply:
ah, do you mean just do `eval.value = s"$funcFullName(${ctx.INPUT_ROW})"`? Let me try

Reply:
I think it would simply become:
```
eval.code = s"${eval.value} = $funcFullName(${ctx.INPUT_ROW});"
```

Reply:
this won't work because `${eval.value}` is not declared if it's not a global variable. I went with
```
eval.code = ""
eval.value = s"$funcFullName(${ctx.INPUT_ROW})"
```

Reply:
I see, sorry, you're right. Then I think your previous solution is better: in this way if `eval.value` is used multiple times we are recomputing the function every time, thus your original implementation was fine, sorry for the bad comment.